cryptoki-rustcrypto/src/rsa/mod.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright 2023 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+use cryptoki::{
+    mechanism::{
+        rsa::{PkcsMgfType, PkcsPssParams},
+        Mechanism, MechanismType,
+    },
+    object::{Attribute, AttributeType, KeyType, ObjectClass},
+};
+use der::oid::AssociatedOid;
+use rsa::{BigUint, RsaPublicKey};
+use signature::digest::Digest;
+use std::convert::TryInto;
+use thiserror::Error;
+
+use crate::SessionLike;
+
+pub mod pkcs1v15;
+pub mod pss;
+
+pub fn read_key<S: SessionLike>(
+    session: &S,
+    template: impl Into<Vec<Attribute>>,
+) -> Result<RsaPublicKey, Error> {
+    let mut template: Vec<Attribute> = template.into();
+    template.push(Attribute::Class(ObjectClass::PUBLIC_KEY));
+    template.push(Attribute::KeyType(KeyType::RSA));
+
+    let keys = session.find_objects(&template)?;
+    if let Some(key) = keys.first() {
+        let attribute_pk = session.get_attributes(
+            *key,
+            &[AttributeType::Modulus, AttributeType::PublicExponent],
+        )?;
+
+        let mut modulus = None;
+        let mut public_exponent = None;
+
+        for attribute in attribute_pk {
+            match attribute {
+                Attribute::Modulus(m) if modulus.is_none() => {
+                    modulus = Some(m.clone());
+                }
+                Attribute::PublicExponent(e) if public_exponent.is_none() => {
+                    public_exponent = Some(e.clone());
+                }
+                _ => {}
+            }
+        }
+
+        let modulus = modulus
+            .ok_or(Error::MissingAttribute(AttributeType::Modulus))
+            .map(|v| BigUint::from_bytes_be(v.as_slice()))?;
+        let public_exponent = public_exponent
+            .ok_or(Error::MissingAttribute(AttributeType::PublicExponent))
+            .map(|v| BigUint::from_bytes_be(v.as_slice()))?;
+
+        Ok(RsaPublicKey::new(modulus, public_exponent)?)
+    } else {
+        Err(Error::MissingKey)
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Cryptoki error: {0}")]
+    Cryptoki(#[from] cryptoki::error::Error),
+
+    #[error("Private key missing attribute: {0}")]
+    MissingAttribute(AttributeType),
+
+    #[error("RSA error: {0}")]
+    Rsa(#[from] rsa::Error),
+
+    #[error("Key not found")]
+    MissingKey,
+}
```

Comment:
I'm curious - why have a different error struct per module? Seems like it could add some headaches if someone wants to use multiple crypto primitives in the same file, they'd have to import and rename them.